src/Context.hs

```diff
@@ -118,6 +119,11 @@ replaceInternalEnv :: Context -> Env -> Context
 replaceInternalEnv ctx env =
   ctx {contextInternalEnv = Just env}
 
+-- The previous environment is completely replaced and will not be recoverable.
+replaceInternalEnvMaybe :: Context -> Maybe Env -> Context
```

Comment:
This name is a bit confusing since it sounds like it will maybe replace the env (but it's always replaced). I don't really have a better suggestion though, I'm afraid...

Reply:
Hopefully the type of the function helps clear up the confusion?